internal/ls/folding.go

```diff
@@ -0,0 +1,541 @@
+	switch n.Kind {
+	case ast.KindBlock:
+		if ast.IsFunctionLike(n.Parent) {
+			return functionSpan(n.Parent, n, sourceFile, l)
+		}
+		// Check if the block is standalone, or 'attached' to some parent statement.
+		// If the latter, we want to collapse the block, but consider its hint span
+		// to be the entire span of the parent.
+		switch n.Parent.Kind {
+		case ast.KindDoStatement, ast.KindForInStatement, ast.KindForOfStatement, ast.KindForStatement, ast.KindIfStatement, ast.KindWhileStatement, ast.KindWithStatement, ast.KindCatchClause:
+			return spanForNode(n, ast.KindOpenBraceToken, true /*useFullStart */, sourceFile, l)
+		case ast.KindTryStatement:
+			// Could be the try-block, or the finally-block.
+			tryStatement := n.Parent.AsTryStatement()
+			if tryStatement.TryBlock == n {
+				return spanForNode(n, ast.KindOpenBraceToken, true /*useFullStart */, sourceFile, l)
+			} else if tryStatement.FinallyBlock == n {
+				node := findChildOfKind(n.Parent, ast.KindFinallyKeyword, sourceFile)
+				if node != nil {
+					return spanForNode(n, ast.KindOpenBraceToken, true /*useFullStart */, sourceFile, l)
+				}
+			}
+		default:
+			// Block was a standalone block.  In this case we want to only collapse
+			// the span of the block, independent of any parent span.
+			return createFoldingRange(l.createLspRangeFromNode(n, sourceFile), "", "")
+		}
+	case ast.KindModuleBlock:
+		return spanForNode(n, ast.KindOpenBraceToken, true /*useFullStart */, sourceFile, l)
+	case ast.KindClassDeclaration, ast.KindClassExpression, ast.KindInterfaceDeclaration, ast.KindEnumDeclaration, ast.KindCaseBlock, ast.KindTypeLiteral, ast.KindObjectBindingPattern:
+		return spanForNode(n, ast.KindOpenBraceToken, true /*useFullStart */, sourceFile, l)
+	case ast.KindTupleType:
+		return spanForNode(n, ast.KindOpenBracketToken, !ast.IsTupleTypeNode(n.Parent) /*useFullStart */, sourceFile, l)
+	case ast.KindCaseClause, ast.KindDefaultClause:
+		return spanForNodeArray(n.AsCaseOrDefaultClause().Statements, sourceFile, l)
+	case ast.KindObjectLiteralExpression:
+		return spanForNode(n, ast.KindOpenBraceToken, !ast.IsArrayLiteralExpression(n.Parent) && !ast.IsCallExpression(n.Parent) /*useFullStart */, sourceFile, l)
+	case ast.KindArrayLiteralExpression:
+		return spanForNode(n, ast.KindOpenBracketToken, !ast.IsArrayLiteralExpression(n.Parent) && !ast.IsCallExpression(n.Parent) /*useFullStart */, sourceFile, l)
+	case ast.KindJsxElement, ast.KindJsxFragment:
+		return spanForJSXElement(n, sourceFile, l)
+	case ast.KindJsxSelfClosingElement, ast.KindJsxOpeningElement:
+		return spanForJSXAttributes(n, sourceFile, l)
+	case ast.KindTemplateExpression, ast.KindNoSubstitutionTemplateLiteral:
+		return spanForTemplateLiteral(n, sourceFile, l)
+	case ast.KindArrayBindingPattern:
+		return spanForNode(n, ast.KindOpenBracketToken, !ast.IsBindingElement(n.Parent) /*useFullStart */, sourceFile, l)
```

Comment:
A lot of these `useFullStart` comments have a trailing space on the end. Can you clean that up?